### ISSUE TITLE
async fn can't handle multiple lifetimes in a slice of slices

### ISSUE THREAD

**Issue by asomers** [rust-lang/rust#76547]
I tried this code:

```rust
use futures::{
    Future,
    task::{Context, Poll}
};
use std::pin::Pin;

pub struct ListFut<'a>(&'a mut [&'a mut [u8]]);

impl<'a> Future for ListFut<'a> {
    type Output = ();

    fn poll(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<Self::Output> {
        unimplemented!()
    }
}

/// The compiler complains that "this parameter and the return type are declared
/// with different lifetimes"
pub async fn readv_at(bufs: &mut [&mut [u8]]) {
    ListFut(bufs).await
}

/// The same thing, but making the lifetimes explicit
pub async fn readv_at2<'a, 'b>(bufs: &'a mut [&'b mut [u8]]) {
    ListFut(bufs).await
}

/// But the compiler accepts this method just fine, because there is only one
/// lifetime.
pub async fn readv_at1<'a>(bufs: &'a mut [&'a mut [u8]]) {
    ListFut(bufs).await
}

/// Even if we explicitly bound one lifetime by the other, the compiler still
/// can't handle it.
pub async fn readv_at3<'a, 'b: 'a>(bufs: &'a mut [&'b mut [u8]]) {
    ListFut(bufs).await
}
```

I expected to see this happen: No error

Instead, this happened: For every variant with two lifetimes, the compiler complains "these two types are declared with different lifetimes... but data from `bufs` flows into `bufs` here"

This looks similar to https://github.com/rust-lang/rust/issues/56238 , but I see that the commit which closed that issue didn't included any test cases involving slices.

### Meta

`rustc --version --verbose`:
```
rustc 1.48.0-nightly (d006f5734 2020-08-28)
rustc 1.42.0 (b8cedc004 2020-03-09)
```

**Comment by Matthias247** [comment]
As discussed on discord I'm not sure whether this is a bug. It works if `ListFut` is declared as 

```rust
pub struct ListFut<'a, 'b>(&'a mut [&'b mut [u8]]);
```

it works. That allows the 2 lifetimes to be separate - compared the original declaration which forces them to be the same. 

It could be that the compiler interprets:

```rust
pub async fn readv_at(bufs: &mut [&mut [u8]]) {
    ListFut(bufs).await
}
```

as

> those 2 lifetimes could be different

However I'm not sure whether this makes a difference here. On the first glance it doesn't, because all those references are pure "inputs". Nothing is returned, so as long as the lifetimes are covering the function duration we should be good - independent whether they are the same or different.

However on the other hand `readv_at` returns a hidden `impl Future`, which carries the input lifetime forwards. Based on that it could matter if these are different or not.

**Comment by taiki-e** [comment]
This is because elided lifetimes are handled as different lifetimes. (in this case, handled as `&'1 mut &'2 mut [u8]`) 

It can be reproduced not only with async functions but also with normal functions:
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=98740b5194107df8a57e41143eebf20e)

```rust
pub struct ListFut<'a>(&'a mut [&'a mut [u8]]);

// error
pub fn readv_at(bufs: &mut [&mut [u8]]) {
    ListFut(bufs);
}

// error
pub fn readv_at2<'a, 'b>(bufs: &'a mut [&'b mut [u8]]) {
    ListFut(bufs);
}

// ok
pub fn readv_at1<'a>(bufs: &'a mut [&'a mut [u8]]) {
    ListFut(bufs);
}

// error
pub fn readv_at3<'a, 'b: 'a>(bufs: &'a mut [&'b mut [u8]]) {
    ListFut(bufs);
}
```

**Comment by taiki-e** [comment]
> ```rust
> /// Even if we explicitly bound one lifetime by the other, the compiler still
> /// can't handle it.
> pub async fn readv_at3<'a, 'b: 'a>(bufs: &'a mut [&'b mut [u8]]) {
>     ListFut(bufs).await
> }
> ```

fwiw, `'b` overlives than `'a`, so you could compile it using a lifetime bound like `'a: 'b, 'b`.

**Comment by tmandry** [comment]
[Discussed] in our async foundations triage meeting today.

We can solve this in two parts.. one would be to change the wording "...but data from `bufs` is returned here" in the async case to make it more clear that we're returning a future (or just not use the special case wording for returned at all.)

The other is not async-related, really, but has to do with improving the error message wording and suggestion.

**Comment by sjakobi** [comment]
Is there still work to do here now that #76765 has landed?

**Comment by halvko** [comment]
On the current nightly:
rustc 1.58.0-nightly (0d1754e8b 2021-11-05)

The sync version of the problem:

```rust
pub struct NestedList<'a>(&'a mut [&'a mut [u8]]);

pub fn readv_at(bufs: &mut [&mut [u8]]) -> NestedList {
    bufs
}
```

The error messages nicely guides the developer to fix their lifetimes, but in the async version no such guidance is given:

```
error[E0623]: lifetime mismatch
  --> src/main.rs:24:13
   |
23 | pub async fn readv_at(bufs: &mut [&mut [u8]]) {
   |                                   ---------   -
   |                                   |           |
   |                                   |           this `async fn` implicitly returns an `impl Future<Output = ()>`
   |                                   this parameter and the returned future are declared with different lifetimes...
24 |     ListFut(bufs).await
   |             ^^^^ ...but data from `bufs` is held across an await point here
```
For an experienced rust developer the first line is somewhat usefull (lifetime mismatch), and the specific await point could be enough of a hint to look at `ListFuts` documentation, but nothing hints at `ListFuts` lifetime requirements.

For an not so experienced rust developer I'd say the error message is less than usefull, as it hints at the need to lifetime annotate the returned future.

**Comment by tmandry** [comment]
@rustbot label: -AsyncAwait-Triaged

Re-submitting to the triage queue since things have changed.

**Comment by nikomatsakis** [comment]
This is probably covered in the above comments, but just a few notes:

The problem here is that you have `&mut [&mut [u8]]` in the function, which is indeed two distinct lifetimes, and only one lifetime in the structure. It doesn't seem to be particularly related to async. The error message is just calibrated for the case of "parameters with simple types" (that have a single lifetime) and not able to cope with this case. Definitely a bug, but a diganostic bug. 

[Playground for a non-async version](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=74388bdf679c7f080303a2014ee75eda) that gets the same error.

**Comment by TornaxO7** [comment]
> On the current nightly: rustc 1.58.0-nightly ([0d1754e](https://github.com/rust-lang/rust/commit/0d1754e8bf6942b4c1d24d7c923438782129ba5a) 2021-11-05)
> 
> The sync version of the problem:
> 
> ```rust
> pub struct NestedList<'a>(&'a mut [&'a mut [u8]]);
> 
> pub fn readv_at(bufs: &mut [&mut [u8]]) -> NestedList {
>     bufs
> }
> ```
> 
> The error messages nicely guides the developer to fix their lifetimes, but in the async version no such guidance is given:
> 
> ```
> error[E0623]: lifetime mismatch
>   --> src/main.rs:24:13
>    |
> 23 | pub async fn readv_at(bufs: &mut [&mut [u8]]) {
>    |                                   ---------   -
>    |                                   |           |
>    |                                   |           this `async fn` implicitly returns an `impl Future<Output = ()>`
>    |                                   this parameter and the returned future are declared with different lifetimes...
> 24 |     ListFut(bufs).await
>    |             ^^^^ ...but data from `bufs` is held across an await point here
> ```
> 
> For an experienced rust developer the first line is somewhat usefull (lifetime mismatch), and the specific await point could be enough of a hint to look at `ListFuts` documentation, but nothing hints at `ListFuts` lifetime requirements.
> 
> For an not so experienced rust developer I'd say the error message is less than usefull, as it hints at the need to lifetime annotate the returned future.

Hi! Some time have passed now and I'm getting for the following file:
```rust
pub struct NestedList<'a>(&'a mut [&'a mut [u8]]);

pub fn readv_at(bufs: &mut [&mut [u8]]) -> NestedList {
    bufs
}

fn main() {}
```
the following output with `cargo check`:
```rust
error[E0106]: missing lifetime specifier
 --> src/main.rs:3:44
  |
3 | pub fn readv_at(bufs: &mut [&mut [u8]]) -> NestedList {
  |                       ----------------     ^^^^^^^^^^ expected named lifetime parameter
  |
  = help: this function's return type contains a borrowed value, but the signature does not say which one of `bufs`'s 2 lifetimes it is borrowed from
help: consider introducing a named lifetime parameter
  |
3 | pub fn readv_at<'a>(bufs: &'a mut [&'a mut [u8]]) -> NestedList<'a> {
  |                ++++        ++       ++                         ++++

error[E0308]: mismatched types
 --> src/main.rs:4:5
  |
3 | pub fn readv_at(bufs: &mut [&mut [u8]]) -> NestedList {
  |                                            ---------- expected `NestedList<'_>` because of return type
4 |     bufs
  |     ^^^^ expected `NestedList<'_>`, found `&mut [&mut [u8]]`
  |
help: try wrapping the expression in `NestedList`
  |
4 |     NestedList(bufs)
  |     +++++++++++    +
```

After applying the suggested fixes, the compiler doesn't complain anymore.

So can we close this issue? @tmandry @halvko

**Comment by tmandry** [comment]
Thanks for noticing @TornaxO7. It looks like this has been fixed, including the suggestions for future work I left in #92183.
